sdk/include/ndk/haltypes.h

```diff
@@ -39,61 +162,1375 @@ typedef enum _FIRMWARE_REENTRY
+
+typedef
+VOID
+(NTAPI *pHalGetEnlightenmentInformation)(
+    _Out_ PHAL_ENLIGHTENMENT_INFORMATION EnlightenmentInformation
+);
+
+typedef
+PVOID
+(NTAPI *pHalAllocateEarlyPages)(
+    _In_ PLOADER_PARAMETER_BLOCK LoaderBlock,
+    _In_ ULONG PageCount,
+    _Out_ PULONG64 PhysicalAddress,
+    _In_ ULONG Protection
+);
+
+typedef
+PVOID
+(NTAPI *pHalMapEarlyPages)(
+    _In_ ULONG64 PhysicalAddress,
+    _In_ ULONG PageCount,
+    _In_ ULONG Protection
+);
+
+#if (NTDDI_VERSION >= NTDDI_WINBLUE)
+typedef
+NTSTATUS
+(NTAPI *pHalPrepareProcessorForIdle)(
+    _In_ ULONG Flags
+);
+
+typedef
+VOID
+(NTAPI *pHalResumeProcessorFromIdle)(
+    VOID
+);
+
+typedef
+VOID
+(NTAPI *pHalNotifyProcessorFreeze)(
+    _In_ BOOLEAN Freezing,
+    _In_ BOOLEAN ThawingToSpinLoop
+);
+#else
+typedef
+NTSTATUS
+(NTAPI *pHalPrepareProcessorForIdle)(
+    ULONG Unknown1,
+    ULONG Unknown2,
+    ULONG Unknown3
+);
+
+typedef
+VOID
+(NTAPI *pHalResumeProcessorFromIdle)(
+    PULONG Unknown
+);
+
+typedef
+VOID
+(NTAPI *pHalNotifyProcessorFreeze)(
+    BOOLEAN Unknown
+);
+#endif
+
+typedef
+VOID
+(NTAPI *PHAL_LOG_ROUTINE)(
+    _In_ ULONG EventId,
+    _In_ PVOID Buffer,
+    _In_ ULONG Size
+);
+
+typedef struct _HAL_LOG_REGISTER_CONTEXT
+{
+    PHAL_LOG_ROUTINE LogRoutine;
+    ULONG Flag;
+} HAL_LOG_REGISTER_CONTEXT, *PHAL_LOG_REGISTER_CONTEXT;
+
+typedef
+VOID
+(NTAPI *pHalRegisterLogRoutine)(
+    _In_ PHAL_LOG_REGISTER_CONTEXT Context
+);
+
+typedef enum _HAL_CLOCK_TIMER_MODE
+{
+    HalClockTimerModePeriodic,
+    HalClockTimerModeOneShot,
+    HalClockTimerModeMax
+} HAL_CLOCK_TIMER_MODE, *PHAL_CLOCK_TIMER_MODE;
+
+typedef struct _HAL_CLOCK_TIMER_CONFIGURATION
+{
+    union {
+        BOOLEAN Flags;
+        struct {
+            BOOLEAN AlwaysOnTimer: 1;
+            BOOLEAN HighLatency: 1;
+            BOOLEAN PerCpuTimer: 1;
+            BOOLEAN DynamicTickSupported: 1;
+        };
+    };
+
+    ULONG KnownType;
+    ULONG Capabilities;
+    ULONG64 MaxIncrement;
+    ULONG MinIncrement;
+} HAL_CLOCK_TIMER_CONFIGURATION, *PHAL_CLOCK_TIMER_CONFIGURATION;
+
+typedef
+ULONG
+(NTAPI *pHalGetClockOwner)(
+    VOID
+);
+
+#if (NTDDI_VERSION >= NTDDI_WINBLUE)
+typedef
+VOID
+(NTAPI *pHalGetClockConfiguration)(
+    _Out_ PHAL_CLOCK_TIMER_CONFIGURATION Configuration
+);
+#else
+typedef
+VOID
+(NTAPI *pHalGetClockConfiguration)(
+    PULONG Unknown1,
+    PULONG Unknown2,
+    PUCHAR Unknown3
+);
+#endif
+
+typedef
+VOID
+(NTAPI *pHalClockTimerActivate)(
+    _In_ BOOLEAN ClockOwner
+);
+
+typedef
+VOID
+(NTAPI *pHalClockTimerInitialize)(
+    VOID
+);
+
+typedef
+VOID
+(NTAPI *pHalClockTimerStop)(
+    VOID
+);
+
+typedef
+NTSTATUS
+(NTAPI *pHalClockTimerArm)(
+    _In_ HAL_CLOCK_TIMER_MODE Mode,
+    _In_ ULONG64 RequestedInteval,
+    _Out_ PULONG64 ActualInterval
+);
+
+typedef
+BOOLEAN
+(NTAPI *pHalTimerOnlyClockInterruptPending)(
+    VOID
+);
+
+typedef struct _HAL_IOMMU_DISPATCH HAL_IOMMU_DISPATCH, *PHAL_IOMMU_DISPATCH;
```

Comment:
Alternatively, do this where the `_HAL_IOMMU_DISPATCH` struct is defined below, and here you could just keep the `struct _HAL_IOMMU_DISPATCH;` only and use `_Inout_ struct _HAL_IOMMU_DISPATCH *DispatchTable` for the callback below...
But I don't have a strong preference, so you decide :)
(It's just that I don't personally like having a typedef and the corresponding struct spread over many places ^^)